src/social_integration.cpp

```diff
@@ -0,0 +1,239 @@
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file social_integration.cpp Base implementation of social integration support. */
+
+#include "stdafx.h"
+
+#include "social_integration.h"
+#include "3rdparty/openttd_social_integration_api/openttd_social_integration_api.h"
+
+#include "debug.h"
+#include "fileio_func.h"
+#include "library_loader.h"
+#include "rev.h"
+#include "string_func.h"
+
+#include "safeguards.h"
+
+/**
+ * Container to track information per plugin.
+ */
+class InternalSocialIntegrationPlugin {
+public:
+	InternalSocialIntegrationPlugin(const std::string &filename, const std::string &basepath) : library(filename), external(basepath)
+	{
+		openttd_info.openttd_version = _openttd_revision;
+	}
+
+	OpenTTD_SocialIntegration_v1_PluginInfo plugin_info = {}; ///< Information supplied by plugin.
+	OpenTTD_SocialIntegration_v1_PluginApi plugin_api = {}; ///< API supplied by plugin.
+	OpenTTD_SocialIntegration_v1_OpenTTDInfo openttd_info = {}; ///< Information supplied by OpenTTD.
+
+	LibraryLoader library; ///< Library handle.
+
+	SocialIntegrationPlugin external; ///< Information of the plugin to be used by other parts of our codebase.
+};
+
+static std::vector<std::unique_ptr<InternalSocialIntegrationPlugin>> _plugins; ///< List of loaded plugins.
+static std::set<std::string> _loaded_social_platform; ///< List of Social Platform plugins already loaded. Used to prevent loading a plugin for the same Social Platform twice.
+static std::vector<SocialIntegrationPlugin *> _external_plugins; ///< List of loaded plugins, but only their "external" part.
+
+static void UpdateExternalPlugins()
+{
+	_external_plugins.clear();
+	for (auto &plugin : _plugins) {
+		_external_plugins.push_back(&plugin->external);
+	}
+}
+
+/** Helper for scanning for files with SocialIntegration as extension */
+class SocialIntegrationFileScanner : FileScanner {
+public:
+	void Scan()
+	{
+#ifdef _WIN32
+		std::string extension = "-social.dll";
+#elif defined(__APPLE__)
+		std::string extension = "-social.dylib";
+#else
+		std::string extension = "-social.so";
+#endif
+
+		this->FileScanner::Scan(extension.c_str(), SOCIAL_INTEGRATION_DIR, false);
+	}
+
+	bool AddFile(const std::string &filename, size_t basepath_length, const std::string &) override
+	{
+		std::string basepath = filename.substr(basepath_length);
+		Debug(misc, 1, "[Social Integration: {}] Loading ...", basepath);
+
+		auto &plugin = _plugins.emplace_back(std::make_unique<InternalSocialIntegrationPlugin>(filename, basepath));
+		UpdateExternalPlugins();
```

Comment:
I would classify this as over-optimizing :) Pretty sure the list will never get long enough to show up on any performance benchmark :) I would rather keep it local to where it is important.

That said, I will see if I can rework it to be a bit nicer. But not for the performance :P More because it is a bit awkward how I did it :)